File: src/main/java/days/Day13.java
package days;

import setup.Day;
import static util.Annotations.*;
import util.Grid;

import java.util.ArrayList;
import java.util.List;

import static util.CollectionUtil.differenceIndices;
import static util.Grid.parseGrid;
import static util.Grid.stripTileData;

public class Day13 extends Day {

    private ArrayList<Grid<String>> grids;

    @Override
    public void processInput() {
        var split = input.split("(\r?\n){2}");
        grids = new ArrayList<>();

        for (var s : split) {
            grids.add(parseGrid(s));
        }
    }

    @Solution("29213")
    @Override
    public Object part1() {
        long sum = 0;

        for (Grid<String> grid : grids) {
            var s = getSymmetryNumber(grid, 0);

            assert s.size() == 1;
            sum += s.get(0);
        }

        return sum;
    }

    private List<Integer> getSymmetryNumber(Grid<String> grid, int allowedErrors) {
        List<Integer> ret = new ArrayList<>();

        for (int i = 0; i < grid.width - 1; i++) {
            if (symmetryErrors(grid, i, false) == allowedErrors) {
                ret.add(i + 1);
            }
        }

        for (int i = 0; i < grid.height - 1; i++) {
            if (symmetryErrors(grid, i, true) == allowedErrors) {
                ret.add(i * 100 + 100);
            }
        }

        return ret;
    }

    private int symmetryErrors(Grid<String> grid, int mirrorIndex, boolean horizontalAxis) {
        int errors = 0;

        for (int i = mirrorIndex; i >= 0; i--) {
            int j = mirrorIndex + mirrorIndex - i + 1;

            if (horizontalAxis) {
                if (j >= grid.height) {
                    continue;
                }
            } else {
                if (j >= grid.width) {
                    continue;
                }
            }

            var r1 = horizontalAxis ? stripTileData(grid.getRow(i)) : stripTileData(grid.getColumn(i));
            var r2 = horizontalAxis ? stripTileData(grid.getRow(j)) : stripTileData(grid.getColumn(j));

            errors += differenceIndices(r1, r2).size();
        }

        return errors;
    }

    @Solution("37453")
    @Override
    public Object part2() {
        long sum = 0;

        for (Grid<String> grid : grids) {
            var s = getSymmetryNumber(grid, 1);

            assert s.size() == 1;
            sum += s.get(0);
        }

        return sum;
    }

    @Override
    public int getDay() {
        return 13;
    }
}

File: src/days/day13.py
from itertools import permutations

from src.setup.day import Day
from src.util.solution import solution

class Day13(Day):
    @property
    def day(self):
        return 13

    @solution("")
    def part1(self) -> object:
        people = set()
        happiness = {}

        for line in self.input:
            line = line.split()
            person = line[0]
            neighbour = line[-1][:-1]
            happiness[(person, neighbour)] = int(line[3]) if line[2] == "gain" else -int(line[3])
            people.add(person)

        max_happiness = 0
        for perm in permutations(people):
            perm = list(perm)
            perm.append(perm[0])
            total = 0
            for i in range(len(perm) - 1):
                total += happiness[(perm[i], perm[i + 1])] + happiness[(perm[i + 1], perm[i])]
            max_happiness = max(max_happiness, total)

        self.people = people
        self.happiness = happiness
        return max_happiness

    @solution("668")
    def part2(self) -> object:
        people, happiness = self.people, self.happiness
        for person in people:
            happiness[("Me", person)] = 0
            happiness[(person, "Me")] = 0
        people.add("Me")

        max_happiness = 0
        for perm in permutations(people):
            perm = list(perm)
            perm.append(perm[0])
            total = 0
            for i in range(len(perm) - 1):
                total += happiness[(perm[i], perm[i + 1])] + happiness[(perm[i + 1], perm[i])]
            max_happiness = max(max_happiness, total)
        return max_happiness

# Day13("test").run()
Day13().run()
